### PR TITLE
Modified quaternion2euler function to cap the arcsin's argument by +-1

### PR DESCRIPTION
**Pull request by Fifourche**
# Description

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

Adresses #2525 .

# How has this been tested?

Checked that returned tuple of angles doesn't contain `nan`, and doesn't return a warning anymore.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: napari/_vispy/quaternion.py
Comment (goanpeca) [inline napari/_vispy/quaternion.py]: Nice method!

What do you think of moving this outside of the function as a private one?

If you agree it would be nice to add a small test for it. Testing inner functions can be hard.

File: napari/_vispy/quaternion.py
Comment (goanpeca) [inline napari/_vispy/quaternion.py]: Could we please add a docstring here? Thanks!

**Review by goanpeca — CHANGES_REQUESTED** [review]
Hi @Fifourche, thanks a lot for working on this 🥳 !

Added some comments on the code! Also to fix the back and isort CI tests please take a look at the `pre-commit install` section of the [Contributing Guide](https://napari.org/developers/contributing.html)

Cheers 😄 !

File: napari/_vispy/quaternion.py
Comment (Fifourche) [reply napari/_vispy/quaternion.py]: Hey, thanks for the nice message ! :) I'm new here, as you might have noticed... I change the file a bit so that no inner function is needed, and it's still readable.

As it's my first time, I'm not used to the customs ; I hope it's still useful !

**Comment by sofroniewn** [comment]
I think this approach of not using the additional capper function looks good. As @goanpeca some tests would be great. It looks like `quaternion2euler` didn't have any of it owns dedicated tests yet, so adding some both for the general functionality and then this specific bug fix would be much appreciated. We can help advise more around the best way to do that too if needed.

Also it looks like our `black` formatting test is failing. There are some instructions on getting `black` working with a pre-commit hook here that you might find helpful https://napari.org/developers/contributing.html#setting-up-a-development-installation.

Thanks again for the PR!!

**Comment by Fifourche** [comment]
Thanks again for your advices ; I feel I'm getting closer to the right track ! I began using `pre-commit` so the code is ok with this regard, and I rewrote a part of the pre-existing unittest for `quaternion2euler`. I made it so that it also check at least one problematic set of angles, so right now the function doesn't pass the test as I raised it in #2525 . I'll work on how to rewrite the quaternion to Euler angles conversion !

By the way, the `quaternion2euler` unittest was in a file named test_utils ; should I rename it to test_quaternion2euler, or was the function meant to be added in the utils file ?

Thanks again for your help !

File: napari/_vispy/_tests/test_utils.py
Comment (sofroniewn) [inline napari/_vispy/_tests/test_utils.py]: you can use `pytest.mark.parametrize` here (see example https://github.com/napari/napari/blob/102a7e8f845893c874d2b86f9371d41130100b89/napari/utils/transforms/_tests/test_transforms.py#L8-L9) to pass the list of `angles` and then you don't need a for loop inside the test :-)

**Review by sofroniewn — COMMENTED** [review]
Nice, looks like this continues to come along very nicely. No need to move the test out of utils, I think it's fine there, I just missed it when I looked yesterday. I added a tip about using making the test parameterized with pytest.

File: napari/_vispy/_tests/test_utils.py
Comment (Fifourche) [reply napari/_vispy/_tests/test_utils.py]: Nice trick ! :) I added it to the test !

**Comment by Fifourche** [comment]
As Euler angles can be degenerate, I changed the test so that the corresponding quaternions are compared.

For a regular user, it means that setting `viewer.camera.angles` and then getting the same attribute might lead to different values (even though it describes the same view), and that is the expected behaviour. Should I specify this somewhere in the documentation ?

Side question : I'm new to contributing and to testing, and I wanted to ask ; here I modified the test according to the expected behaviour of `quaternion2euler`, but the function itself wouldn't pass this test in its actual state. Will it bother the community of contributors if I leave it like that while I modify the `quaternion2euler` function ? Thanks again for your advices, it's helping me a lot ! :)

File: napari/_vispy/_tests/test_utils.py
Comment (sofroniewn) [inline napari/_vispy/_tests/test_utils.py]: Let's remove this code block and make this a doubly parametrized test as the only thing different here is degrees `True`/ `False`. You can do that by passing a list of 2-tuples `(angle, degrees)` and then pass in 6 values instead of three, see https://docs.pytest.org/en/stable/parametrize.html

**Comment by sofroniewn** [comment]
> For a regular user, it means that setting viewer.camera.angles and then getting the same attribute might lead to different values (even though it describes the same view), and that is the expected behaviour. Should I specify this somewhere in the documentation ?

Yeah, I'd add that to the documentation, certainly to the `quaternion2euler` doc string, and possibly to the `viewer.camera.angles` doc string too.

> Side question : I'm new to contributing and to testing, and I wanted to ask ; here I modified the test according to the expected behaviour of quaternion2euler, but the function itself wouldn't pass this test in its actual state. Will it bother the community of contributors if I leave it like that while I modify the quaternion2euler function ? Thanks again for your advices, it's helping me a lot ! :)

Good question! It's fine to have broken tests in this PR as you do your development, but it will be important to get all the tests passing this PR before we merge it. So either we'll have to figure out a way to update the old tests, or the function to get things working with the new behavior. Having all the tests passing before merge is important as then when others are developing they know they are building on a clean slate. Let me know if you have additional questions!!

**Comment by Fifourche** [comment]
I've modified `quaternion2euler` to take into account the singularity, simplified the test as suggested (I think) and added some docstrings info ! Now tests are passing :)

File: napari/_vispy/_tests/test_utils.py
Comment (sofroniewn) [inline napari/_vispy/_tests/test_utils.py]:
```suggestion
@pytest.mark.parametrize('angles,degrees', angles_param)
def test_quaternion2euler(angles, degress):
```
```suggestion
@pytest.mark.parametrize('angles_param', angles_param)
def test_quaternion2euler(angles_param):
```

If you look here https://docs.pytest.org/en/stable/parametrize.html you can actually unpack the `angles_param` before passing to the test function. This will make everything a little easier to read

File: napari/_vispy/_tests/test_utils.py
Comment (sofroniewn) [inline napari/_vispy/_tests/test_utils.py]:
```suggestion
    q = Quaternion.create_from_euler_angles(*angles, degrees)
```
```suggestion
    q = Quaternion.create_from_euler_angles(*angles_param[0], angles_param[1])
```

File: napari/_vispy/_tests/test_utils.py
Comment (sofroniewn) [inline napari/_vispy/_tests/test_utils.py]:
```suggestion
    ea = quaternion2euler(q, degrees=degrees)
    q_p = Quaternion.create_from_euler_angles(*ea, degrees=degrees)
```
```suggestion
    ea = quaternion2euler(q, degrees=angles_param[1])
    q_p = Quaternion.create_from_euler_angles(*ea, degrees=angles_param[1])
```

**Review by sofroniewn — COMMENTED** [review]
Ok nice, this basically looks done. I've made one additional request to unpack the args for the pytest parametrized function, but then I think this will be ready to approve and merge!! Thanks for your patience and all the work here!!

**Review by sofroniewn — APPROVED** [review]
Great, looks like this is all done! I'm approving now. Will leave it open till tomorrow in case anyone else has any additional comments. Will merge then.

Thanks again for your contribution!

File: napari/_vispy/quaternion.py
Comment (Carreau) [inline napari/_vispy/quaternion.py]:
```suggestion
    that are closer than the chosen epsilon from it.
```

Just removing a leading space. I think that's fine with it, but it may be RST will parse that as a nested block.

**Review by Carreau — APPROVED** [review]

**Pull request merged**
